plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XLanguageServerImpl.java

```diff
@@ -471,132 +455,44 @@ public WorkspaceService getWorkspaceService() {
 
 	@Override
 	public void didOpen(DidOpenTextDocumentParams params) {
-		runBuildable("didOpen", () -> toBuildable(params));
-	}
-
-	/**
-	 * Evaluate the params and deduce the respective build command.
-	 */
-	protected XBuildable toBuildable(DidOpenTextDocumentParams params) {
 		TextDocumentItem textDocument = params.getTextDocument();
-		return workspaceManager.didOpen(getURI(textDocument),
-				textDocument.getVersion(), textDocument.getText());
+		openFilesManager.openFile(getURI(textDocument), textDocument.getVersion(), textDocument.getText());
 	}
 
 	@Override
 	public void didChange(DidChangeTextDocumentParams params) {
-		if (isSourceFileOrOpen(uriExtensions.toUri(params.getTextDocument().getUri()))) {
-			runBuildable("didChange", () -> toBuildable(params));
-		}
-	}
-
-	private boolean isSourceFileOrOpen(URI uri) {
-		if (workspaceManager.isDocumentOpen(uri)) {
-			return true;
-		}
-		return isSourceFile(uri);
-	}
-
-	private boolean isSourceFile(URI uri) {
-		IProjectConfig projectConfig = workspaceManager.getWorkspaceConfig().findProjectContaining(uri);
-		if (projectConfig != null) {
-			ISourceFolder sourceFolder = projectConfig.findSourceFolderContaining(uri);
-			if (sourceFolder != null) {
-				return true;
-			}
-		}
-		return false;
-	}
-
-	/**
-	 * Evaluate the params and deduce the respective build command.
-	 */
-	protected XBuildable toBuildable(DidChangeTextDocumentParams params) {
 		VersionedTextDocumentIdentifier textDocument = params.getTextDocument();
-		return workspaceManager.didChangeTextDocumentContent(getURI(textDocument),
-				textDocument.getVersion(), params.getContentChanges());
+		URI uri = getURI(textDocument);
+		if (openFilesManager.isOpen(uri)) { // FIXME GH-1774 reconsider
```

Comment:
I think this is what the TODO is about, but I'll put my thoughts here nevertheless:
I'm not sure what the LSP client sends to the server if a file is changed on disk and not through the client. We must ensure that we don't miss these kinds of changes.

Reply:
In that case the client sends the separate LSP event `didChangeWatchedFiles`. The `OpenFilesManager` ignores these events entirely. Instead, they are sent to the `LSPBuilder`, triggering an incremental build which will eventually lead to an update of the persisted state index, which will then be forwarded to the `OpenFilesManager`.